spyder/plugins/toolbar/tests/test_toolbar.py

```diff
@@ -0,0 +1,153 @@
+# (see spyder/__init__.py for details)
+
+from unittest.mock import MagicMock
+
+from qtpy.QtWidgets import QMainWindow
+import pytest
+
+from spyder.config.manager import CONF
+from spyder.plugins.toolbar.api import ApplicationToolbars
+from spyder.plugins.toolbar.plugin import Toolbar
+from spyder.plugins.workingdirectory.plugin import WorkingDirectory
+
+
+class MainWindow(QMainWindow):
+
+    _cli_options = MagicMock()
+
+    def get_plugin(self, name, error=True):
+        return MagicMock()
+
+
+@pytest.fixture
+def toolbar(qtbot):
+    # Create toolbar plugin
+    main_window = MainWindow()
+    toolbar = Toolbar(main_window, configuration=CONF)
+    toolbar.on_initialize()
+
+    # Add working directory toolbar
+    cwd = WorkingDirectory(main_window, None)
+    cwd.on_initialize()
+    toolbar.add_application_toolbar(cwd.get_container().toolbar)
+
+    # Add buttons to the other toolbars
+    actions = [
+        (ApplicationToolbars.File, "filenew"),
+        (ApplicationToolbars.Run, "run"),
+        (ApplicationToolbars.Debug, "debug"),
+        (ApplicationToolbars.Main, "python"),
+    ]
+
+    for i, item in enumerate(actions):
+        toolbar.create_action(
+            f"action_{i}",
+            text="Action {i}",
+            icon=toolbar.create_icon(item[1]),
+            triggered=lambda: True,
+        )
+
+        toolbar.add_item_to_application_toolbar(
+            toolbar.get_action(f"action_{i}"),
+            toolbar_id=item[0],
+        )
+
+    # Show toolbars and main window
+    toolbar.on_mainwindow_visible()
+    toolbar.toggle_lock(True)
+    main_window.show()
+    return toolbar
+
+
+def test_default_order(toolbar):
+    """Check toolbars are displayed in their default order."""
+    expected_order = [
+        ApplicationToolbars.File,
+        ApplicationToolbars.Run,
+        ApplicationToolbars.Debug,
+        ApplicationToolbars.Main,
+        ApplicationToolbars.WorkingDirectory,
+    ]
+
+    # Check the horizontal order obtained the displayed toolbar doesn't change
+    # after sorting it.
+    current_order = [
+        toolbar.get_application_toolbar(id_).x()
+        for id_ in expected_order
+    ]
+    assert current_order == sorted(current_order)
+
+
+def test_restore_toolbars_order(toolbar, qtbot):
+    """
+    Check that if a toolbar is moved by users to a different position, then the
+    new order is restored during the next startup.
+    """
+    new_order = [
+        ApplicationToolbars.Debug,
+        ApplicationToolbars.File,
+        ApplicationToolbars.Run,
+        ApplicationToolbars.Main,
+    ]
+
+    # Select two toolbars to change their order
+    file_toolbar = toolbar.get_application_toolbar(ApplicationToolbars.File)
+    debug_toolbar = toolbar.get_application_toolbar(ApplicationToolbars.Debug)
+
+    # Move the debug toolbar to be before the file one
+    toolbar.get_main().insertToolBar(file_toolbar, debug_toolbar)
+    qtbot.wait(500)
```

Comment:
Maybe for some of these `qtbot.wait` calls over these new tests could be replaced for a `qtbot.waitSignal` checking for the new toolbars `sig_is_rendered` signal?

Reply:
`sig_is_rendered` is only called when the `render` method of `SpyderToolbar` or subclasses is called. Since moving a toolbar doesn't render it again, then we can't use that signal here.

By the way, I had to introduce those small waits to make the tests pass, otherwise they fail with odd errors. But in real life such waits are guaranteed because people have to move the mouse or press a keyboard shortcut to do another operation in Spyder.